### ISSUE TITLE
ncurses support

### ISSUE THREAD

**Issue by marrub--** [Team-Doominati/Doominati#31]
**WHY NOT**?

**Comment by marrub--** [comment]
Repurposing this issue.
We should totally have a cool console thing that uses ncurses and displays a neat startup log like Doom.

**Comment by Kyle873** [comment]
As for providing an in-game console, you think we should do that in-engine or leave it for the scripting end?

**Comment by marrub--** [comment]
It should be done in-engine (David's module plan puts it under a module named "Sh") and be mutable from scripts (ie. natives or some other fun stuff to add cvars/ccmds from structures).

**Comment by Kyle873** [comment]
Nice. GOTTA HAVE MUH FANCY CONSOLE WITH ANIMATIONS AND COLORS AND FLASHY TEXT AND BLIP BLOOPS.

**Comment by DavidPH** [comment]
Yeah, it should be possible to implement custom console commands. I have given some thoughts to an API, but obviously the first step is having builtin console commands (which will have the amazing power of gdcc-option-lib at their disposal). Also if it's worthwhile to have an intermediate (possibly long-term as an option) console that just uses stdin and stdout, ie operating in the OS's terminal.

**Comment by marrub--** [comment]
All solutions will likely have to be platform-dependent (due to input blocking) but they'll all be implemented pretty much the same. I want to use PDCursesWin32A (optionally) for a Windows console, and ncurses otherwise, or if that isn't found then just stdio. Of course, if your platform doesn't have non-blocking input, you're just screwed anyway.

**Comment by DavidPH** [comment]
You can implement non-blocking stdin with a thread trivially.

**Comment by marrub--** [comment]
Wow! I didn't even fucking know that. That is way simpler than I thought.

**Comment by DavidPH** [comment]
Almost a year later, and I'm thinking the builtin console should be regarded as primarily a dev/cheat tool that just runs on stdio. No need for anything fancy, just parse and execute commands in the OS terminal. (Including script-defined commands, naturally.) It's simple and not prone to script bugs removing the console as an emergency control interface.

Instead, there would be natives for helping with in-game consoles, both the traditional type and any that are actual game elements. This could take the form of creating and manipulating Console objects each with their own buffer (either line-oriented or ncurses-style) and input. Would be possible to then do rendering as a single `DGE_Draw_Console` call of some sort, potentially. Though of course the buffer would be read/write.

There would not be natives which pass data to the builtin console. This lesson has been learned, already. Though any particularly useful builtin console commands could of course be exposed as natives.

If this seems like a reasonable design, I would propose this issue be narrowed to the former, with a new issue created for the latter.
